pilot: fix Gateway API conversion MatchExpressions handling for NotIn and DoesNotExist operators

mikemorris opened the pull request:
**Please provide a description of this PR:**

Fixes #48044 

Added a few additional listeners in the test fixture YAML to add coverage for AllowedRoutes selector MatchExpressions, but think I missed updating a bit of boilerplate to get the golden files regenerating properly (and I might need to add additional routes to target these new listeners?) - I'm currently hitting the following error when running `REFRESH_GOLDEN=1 go test ./pilot/pkg/config/kube/gateway -run TestConvertResources/route-binding -v`

```
    diff.go:111: Refreshing golden file testdata/route-binding.yaml.golden
    conversion_test.go:1229: failed to parse CRD: configuration is invalid: 2 errors occurred:
                * config namespace and dnsName in host entry cannot be empty
                * invalid namespace value "" in gateway
```

mikemorris committed:
pilot: track all namespaces when using NotIn or DoesNotExist selector MatchExpressions

mikemorris committed:
pilot: add test cases for selector MatchExpressions route binding

mikemorris committed:
pilot: update golden files for TestConvertResources/route-binding

istio-testing commented:
Hi @mikemorris. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

howardjohn commented on pilot/pkg/config/kube/gateway/conversion.go:
```diff
@@ -2247,6 +2247,10 @@ func getNamespaceLabelReferences(routes *k8s.AllowedRoutes) []string {
 		res = append(res, k)
 	}
 	for _, me := range routes.Namespaces.Selector.MatchExpressions {
+		if me.Operator == "NotIn" || me.Operator == "DoesNotExist" {
```
nit: use LabelSelectorOpDoesNotExist from metav1

howardjohn reviewed (approved):
LGTM. A comment along the lines of "over-matching is fine, we just cannot under match" would help explain this. But looks good

keithmattix commented:
/ok-to-test
/hold until John's feedback can be addressed.

istio-testing commented:
@mikemorris: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1729628481492357120) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1729628481492357120) | true | `/test release-notes`
unit-tests-arm64_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests-arm64_istio/1729628468162859008) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1729628481492357120) | true | `/test release-notes`
unit-tests-arm64_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests-arm64_istio/1729628468162859008) | true | `/test unit-tests-arm64`
unit-tests_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests_istio/1729628468108333056) | true | `/test unit-tests`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

MorrisLaw commented on pilot/pkg/config/kube/gateway/testdata/route-binding.status.yaml.golden:
```diff
@@ -247,7 +247,7 @@ status:
       name: gateway
       namespace: istio-system
-      sectionName: namespace-selector
+      sectionName: selector-labels	
```
nit: looks like extra whitespace on this line, right after labels

mikemorris committed:
wip: backtrack to working set of tests

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests-arm64_istio/1729628468162859008) | true | `/test unit-tests-arm64`
unit-tests_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests_istio/1729628468108333056) | true | `/test unit-tests`
release-notes_istio | 2fcebf6b24268a92807a5467150085ab1f2e61c1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1732295171560706048) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 9d7ab74a57374fd103d7d4881f765448d564c82d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests-arm64_istio/1729628468162859008) | true | `/test unit-tests-arm64`
release-notes_istio | 5ad840f6ba82214ee9a6348a7c141263b9227639 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1732295793483714560) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mikemorris committed:
wip: enable expr-in and selector-combined test cases

istio-testing commented:
@mikemorris: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 1caa89cee95438b42d0bb2f8f53fc14eb3988e34 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1732305973483147264) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 1caa89cee95438b42d0bb2f8f53fc14eb3988e34 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1732305973483147264) | true | `/test release-notes`
lint_istio | 1caa89cee95438b42d0bb2f8f53fc14eb3988e34 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/lint_istio/1732305971813814272) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 1caa89cee95438b42d0bb2f8f53fc14eb3988e34 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1732305973483147264) | true | `/test release-notes`
lint_istio | 1caa89cee95438b42d0bb2f8f53fc14eb3988e34 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/lint_istio/1732305971813814272) | true | `/test lint`
gencheck_istio | 1caa89cee95438b42d0bb2f8f53fc14eb3988e34 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/gencheck_istio/1732305972640092160) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mikemorris commented on pilot/pkg/config/kube/gateway/conversion.go:
```diff
@@ -2247,6 +2247,10 @@ func getNamespaceLabelReferences(routes *k8s.AllowedRoutes) []string {
 	}
 	for _, me := range routes.Namespaces.Selector.MatchExpressions {
+		if me.Operator == "NotIn" || me.Operator == "DoesNotExist" {
+			res = append(res, "*")
```
The `.Intersection` method inside `namespaceEvent` implements this.

https://github.com/istio/istio/blob/64d8e69aa62e27068312f9b7006450aaaa4aa336/pilot/pkg/config/kube/gateway/controller.go#L312-L335

mikemorris replied:
https://github.com/istio/istio/blob/master/pkg/config/host/names.go#L75-L99

mikemorris committed:
pilot: exclude empty string ns when building Istio Gateway hosts from Gateway API resource

mikemorris committed:
fixup: revert debugging snippet

mikemorris committed:
pilot: update comments for Gateway API AllowedRoutes matching

mikemorris committed:
add release note

istio-testing commented:
@mikemorris: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 7919302b347a28980ed5fb4243ce426e832abe59 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/release-notes_istio/1732477753460002816) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mikemorris commented on pilot/pkg/config/kube/gateway/conversion.go:
```diff
@@ -2414,7 +2423,7 @@ func parentRefString(ref k8s.ParentReference) string {
 }
 
-// buildHostnameMatch generates a VirtualService.spec.hosts section from a listener
+// buildHostnameMatch generates a Gateway.spec.servers.hosts section from a listener
```
This comment appears to have been inaccurate - the VirtualService.spec.hosts section would be built from HTTPRoute.spec.hostnames - [Listener](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.Listener) configuration is used to build the Istio Gateway servers.

mikemorris commented on releasenotes/notes/gateway-allowedroutes-fix.yaml:
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: networking
```
```suggestion
area: traffic-management
```

mikemorris committed:
fix release note area category

mikemorris commented on pilot/pkg/config/kube/gateway/conversion.go:
```diff
@@ -2424,7 +2433,10 @@ func buildHostnameMatch(localNamespace string, r GatewayResources, l k8s.Listene
 	resp := []string{}
 	for _, ns := range namespacesFromSelector(localNamespace, r, l.AllowedRoutes) {
-		resp = append(resp, fmt.Sprintf("%s/%s", ns, hostname))
+		// This check is necessary to prevent adding a hostname with an invalid empty namespace
+		if len(ns) > 0 {
+			resp = append(resp, fmt.Sprintf("%s/%s", ns, hostname))
+		}
```
Excluding the `""` empty string namespace picked up by tracking `"*"` when building the Istio Gateway server hosts array was the trick needed to fix the invalid generated CRDs.

istio-testing commented:
@mikemorris: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-basic-arm64_istio | e46a62df82104afdf4006775c344e690554cd7cc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/integ-basic-arm64_istio/1732479234351304704) | true | `/test integ-basic-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mikemorris committed:
expand test coverage

mikemorris committed:
validation: allow ~ in gateway slash wildcard hosts

This is used when converting from Gateway API gateways and no
matching routes can be found, because hosts must be non-empty.

mikemorris commented on pkg/config/validation/validation.go:
```diff
@@ -1075,19 +1075,10 @@ func validateNamespaceSlashWildcardHostname(hostname string, isGateway bool) (er
-			if !labels.IsDNS1123Label(parts[0]) {
-				errs = appendErrors(errs, fmt.Errorf("invalid namespace value %q in gateway", parts[0]))
-			}
+	// namespace can be * or . or ~ or a valid DNS label in sidecars or gateways
+	if parts[0] != "*" && parts[0] != "." && parts[0] != "~" {
+		if !labels.IsDNS1123Label(parts[0]) {
+			errs = appendErrors(errs, fmt.Errorf("invalid namespace value %q", parts[0]))
```
`~` namespace labels _are_ used by Istio Gateways when converting from Gateway API gateways and no matching routes can be found - I believe these generated resources were incorrectly failing validation before, see https://github.com/istio/istio/blob/e46a62df82104afdf4006775c344e690554cd7cc/pilot/pkg/config/kube/gateway/conversion.go#L2442-L2447

howardjohn replied:
The internal generate ones don't actually go through validation in real world -- only in the tests. 

Not sure how much we care about maintaining that property. Looks like in other parts we allowlist these "internal gateways" (see code like `gatewaySemantics := cfg.Annotations[constants.InternalRouteSemantics] == constants.RouteSemanticsGateway`)

mikemorris replied:
Rewrote this logic to plumb through a new `internal.istio.io/gateway-semantics` annotation, see latest commit.

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-basic-arm64_istio | e46a62df82104afdf4006775c344e690554cd7cc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/integ-basic-arm64_istio/1732479234351304704) | true | `/test integ-basic-arm64`
unit-tests-arm64_istio | e11d5fbe3010eac19a0ff242e17701f7bb03d57e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests-arm64_istio/1732512580477390848) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mikemorris commented on pilot/pkg/config/kube/gateway/testdata/route-binding.status.yaml.golden:
```diff
@@ -164,7 +164,287 @@ status:
+      kind: HTTPRoute
+    - group: gateway.networking.k8s.io
+      kind: GRPCRoute
+  - attachedRoutes: 1
+    conditions:
+    - lastTransitionTime: fake
+      message: No errors found
+      reason: Accepted
+      status: "True"
+      type: Accepted
+    - lastTransitionTime: fake
+      message: No errors found
+      reason: NoConflicts
+      status: "False"
+      type: Conflicted
+    - lastTransitionTime: fake
+      message: No errors found
+      reason: Programmed
+      status: "True"
+      type: Programmed
+    - lastTransitionTime: fake
+      message: No errors found
+      reason: ResolvedRefs
+      status: "True"
+      type: ResolvedRefs
+    name: slctr-expr-notin-no
```
This is okay, the `group-namespace1` route is excluded but the `default/bind-all` route is not.

mikemorris commented:
I ended up touching a few additional spots of logic that seemed incorrect (left comments highlighting them), this should be complete now but could use another set of eyes.

/retest

mikemorris committed:
add internal.istio.io/gateway-semantics annotation

To track whether an internally-generated config originated from a k8s
gateway-api resource.

istio-testing commented:
@mikemorris: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 8090aa6b13326cf26b26a0a9df6014f16347ec38 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests-arm64_istio/1733205784768024576) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 8090aa6b13326cf26b26a0a9df6014f16347ec38 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests-arm64_istio/1733205784768024576) | true | `/test unit-tests-arm64`
gencheck_istio | 8090aa6b13326cf26b26a0a9df6014f16347ec38 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/gencheck_istio/1733205798311432192) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@mikemorris: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 8090aa6b13326cf26b26a0a9df6014f16347ec38 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/unit-tests-arm64_istio/1733205784768024576) | true | `/test unit-tests-arm64`
gencheck_istio | 8090aa6b13326cf26b26a0a9df6014f16347ec38 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/gencheck_istio/1733205798311432192) | true | `/test gencheck`
lint_istio | 8090aa6b13326cf26b26a0a9df6014f16347ec38 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/48070/lint_istio/1733205797053140992) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mikemorris committed:
add test coverage for Gateway ~/hostname validation, fix logic

dhawton commented on releasenotes/notes/gateway-allowedroutes-fix.yaml:
```diff
@@ -0,0 +1,9 @@
+
+releaseNotes:
+- |
+  **Fixed** Gateway API AllowedRoutes handling for NotIn and DoesNotExist LabelSelector MatchExpressions.
```
```suggestion
  **Fixed** Gateway API `AllowedRoutes` handling for `NotIn` and `DoesNotExist` label selector match expressions.
```

mikemorris committed:
Update releasenotes/notes/gateway-allowedroutes-fix.yaml

Co-authored-by: Daniel Hawton <daniel@hawton.org>
